### PR TITLE
Functionality port from CLI tool + test utility

### PR DESCRIPTION
**Pull request by rionastokes**
Porting over all functionalities from the senecajs/repo-maintain CLI tool, and some housekeeping

### CODE REVIEW

File: checks.js [hunk 1]
```diff
@@ -0,0 +1,85 @@
+module.exports = {
```
Comment (rjrodger) [inline checks.js:1]: formatting: use https://prettier.io/
follow config from: https://github.com/senecajs/seneca/blob/master/package.json
examples: 2 space indent, space after colon

File: maintain.js [hunk 2]
```diff
@@ -1,25 +1,345 @@
-/* Copyright © 2021 Ríona Stokes and other contributors, MIT License. */
-'use strict'
+module.exports = () => {
```
Comment (rjrodger) [inline maintain.js]: explicitly name the function: `Maintain`

File: maintain.js [hunk 3]
```diff
@@ -1,25 +1,345 @@
...
 
-const Fs = require('fs')
+  const Filehound = require('filehound')
+  const Path = require('path')
```
Comment (rjrodger) [inline maintain.js]: separate built-in modules and external modules - easier to see external deps
https://github.com/senecajs/seneca/blob/master/seneca.js#L5

File: maintain.js [hunk 4]
```diff
@@ -1,25 +1,345 @@
...
+  const Fs = require('fs')
+  const Hoek = require('@hapi/hoek')
+  const Marked = require('marked')
 
+  const checkList = require('./checks')
```
Comment (rjrodger) [inline maintain.js]: prefer import style:

```
const { checkList } = require('./checks')
```

```
function checkList { ... }
module.exports = {
  checkList
}
```

File: maintain.js [hunk 5]
```diff
@@ -1,25 +1,345 @@
...
+  const checkOps = checkOperations()
```
Comment (rjrodger) [inline maintain.js]: rename to `defineChecks` as this is just settings things up

File: maintain.js [hunk 6]
```diff
@@ -1,25 +1,345 @@
...
 
-async function Maintain() {
-  console.log('MAINTAIN 0.0.2')
+  async function configDef(){
```
Comment (rjrodger) [inline maintain.js]: put utils at the end of the file

File: maintain.js [hunk 7]
```diff
@@ -1,25 +1,345 @@
...
+    var argString = process.argv.slice(2)
```
Comment (rjrodger) [inline maintain.js]: prefer let and const

File: maintain.js [hunk 8]
```diff
@@ -1,25 +1,345 @@
...
+    if (null == argString[0]) {
+      argString[0] = "base"
+    }
+    const argArray = argString[0].split(',')
+    return argArray
+  }
+
+  async function runChecksPrep(config){
```
Comment (rjrodger) [inline maintain.js]: place further down the file

File: maintain.js [hunk 9]
```diff
@@ -1,25 +1,345 @@
...
+      }
+    }
+    // console.log(relCheckList)
+    return {
+      relCheckList: relCheckList,
+      dataForChecks: dataForChecks
+    }
+  }
+
+  async function runChecks(config){
```
Comment (rjrodger) [inline maintain.js]: make this the first function

File: maintain.js [hunk 10]
```diff
@@ -1,25 +1,345 @@
...
+
+    let prep = await runChecksPrep(config)
+    let relCheckList = prep.relCheckList //ok
+    let dataForChecks = prep.dataForChecks //ok
+    let results = {}
+
+    for(const checkName in relCheckList) {
```
Comment (rjrodger) [inline maintain.js]: lovely!

File: maintain.js [hunk 11]
```diff
@@ -1,25 +1,345 @@
...
+    fails = fails.join('\n\t')
+    let message = `Total checks for this configuration: ${totalNb}
+    \nFailed checks: ${failNb}\n\t${fails}
+    \n${note}`
+    return message
+
+  }
+
+  // --------------------------------------------------------------------
+  async function runAll() {
```
Comment (rjrodger) [inline maintain.js]: seneca-maintain is both a command line tool, and a test utility
command line stuff should be .bin/foo as per repo-maintain

**Review by rjrodger — CHANGES_REQUESTED** [review]

File: checks.js (hunk 1 above)
Comment (rionastokes) [reply checks.js:1]: Resolved with commit d89dfeace145e064fabc2302d04f27258bc65513

File: maintain.js (hunk 2 above)
Comment (rionastokes) [reply maintain.js]: Resolved with commit d89dfeace145e064fabc2302d04f27258bc65513

File: maintain.js (hunk 3 above)
Comment (rionastokes) [reply maintain.js]: Resolved with commit 8ffd82fb6a786b28be7c2e95fd85e53de929fd9c

File: maintain.js (hunk 4 above)
Comment (rionastokes) [reply maintain.js]: Modification to require resolved with commit 8ffd82fb6a786b28be7c2e95fd85e53de929fd9c

checks.js refactored to account for this change with commit 3ea4a65b59e26a928ffa8921e580288b03cd9e50

File: maintain.js (hunk 5 above)
Comment (rionastokes) [reply maintain.js]: Resolved with commit 8ffd82fb6a786b28be7c2e95fd85e53de929fd9c

File: maintain.js (hunk 6 above)
Comment (rionastokes) [reply maintain.js]: Resolved with commit 8ffd82fb6a786b28be7c2e95fd85e53de929fd9c, may still need moving around

File: maintain.js (hunk 7 above)
Comment (rionastokes) [reply maintain.js]: Resolved with commit 8ffd82fb6a786b28be7c2e95fd85e53de929fd9c

File: maintain.js (hunk 8 above)
Comment (rionastokes) [reply maintain.js]: Resolved with commit 8ffd82fb6a786b28be7c2e95fd85e53de929fd9c, may need moving

File: maintain.js (hunk 9 above)
Comment (rionastokes) [reply maintain.js]: Resolved with commit 8ffd82fb6a786b28be7c2e95fd85e53de929fd9c

File: maintain.js (hunk 11 above)
Comment (rionastokes) [reply maintain.js]: Refactoring/restructuring for this currently in progress - PR has been converted to draft in the meantime. Will reopen for review once refactoring is complete

File: import.test.js [hunk 12]
```diff
@@ -0,0 +1,5 @@
+'use strict'
+
+const { Maintain } = require('./newMaintain.js')
```
Comment (rjrodger) [inline import.test.js]: avoid names like "newX" or "index2.html" - they have no semantic content

**Comment by rjrodger** [comment]
remove package-lock.json as this is a library

File: package.json [hunk 13]
```diff
@@ -2,9 +2,13 @@
   "name": "@seneca/maintain",
   "version": "0.0.2",
   "description": "Run maintenance tests for Seneca plugins.",
-  "main": "maintain.js",
+  "main": "run-maintain.js",
```
Comment (rjrodger) [inline package.json]: convention: name of the entry point is the same as the library - `maintain.js`

File: package.json [hunk 14]
```diff
@@ -2,9 +2,13 @@
...
   "scripts": {
-    "test": "echo \"Error: no test specified\" && exit 1"
+    "test": "jest",
+    "prettier": "prettier --write *.js lib/*.js test/*.js"
+  },
+  "bin": {
+    "seneca-maintain": "./run-maintain.js"
```
Comment (rjrodger) [inline package.json]: move the cli script to `bin/cli-maintain.js` - more obvious what it does

File: run-maintain.js [hunk 15]
```diff
@@ -0,0 +1,3 @@
+#!/usr/bin/env node
+const { Maintain } = require('./maintain.js')
+Maintain()
```
Comment (rjrodger) [inline run-maintain.js]: at this point, you know you're running from the cli

```
Maintain({...config...})
```

```
config = {
  cli: true
}
```

also a good place to get `process.argv`

maintain.js should just assume it's pure code

**Comment by rjrodger** [comment]
maintain.js should not assume it is cli

move to bin: https://github.com/senecajs/seneca-maintain/blob/funct-port/maintain.js#L108

**Review by rjrodger — CHANGES_REQUESTED** [review]

**Review by rjrodger — CHANGES_REQUESTED** [review]
update

**Pull request merged**